feat: Adds code modifiers

LukeHankey opened the pull request:
Current functionality:
* remove_indentation
* remove_end_colon
* change_keyword
* comment
* change_function_call_name
* insert_empty_statements

More to come from #5. Structure will vary depending on how we go about transferring the data across the websockets.

LukeHankey committed:
feat: Adds two simple code modifiers

* remove_indentation and remove_end_colon

LukeHankey committed:
docs: consistency between modifier docs

kappq reviewed (commented):
Currently, this removes the indentation/colons wherever possible. This could get a bit annoying to fix in large files, maybe we should make sure that the bug is only introduced at a random point in the code?

LukeHankey committed:
feat: Returns a number of modifiers based on the difficulty

LukeHankey committed:
feat: Add difficulty option

LukeHankey commented:
> Currently, this removes the indentation/colons wherever possible. This could get a bit annoying to fix in large files, maybe we should make sure that the bug is only introduced at a random point in the code?

That's a good idea. I've just pushed two other commits without reading your comment. The new commits now partially solve that with taking only a specific amount of the code, rather than removing every occurrence. Could look into it further to randomize this selection as it currently removes only the first x occurrences.

LukeHankey committed:
feat: adds random sampling to all modifiers

LukeHankey committed:
feat: Add change_keyword modifier

* docs: Update class __doc__ to include about random sampling

LukeHankey committed:
fix: Subsequent method calls now uses the content from previously modified calls and not the original data

* Minimum between the difficulty level and the number of lines (depending on the method) as having a higher difficulty than the number of lines raises a ValueError

LukeHankey committed:
feat: Add comment modifier

LukeHankey committed:
fix: D101 Missing docstring in public class

Using the Google docstring format, you can either pass in the class docstring as the public class or in the __init__ method. Have chosen to use the __init__ method to document the class

LukeHankey committed:
docs: Use google docstring format

Commit:
Added the fastapi dependency requirement (#6)

* Added the fastapi dependency requirement

* Added uvicorn dependancy

kappq reviewed (approved):
Works great!

kappq committed:
Add pep8-naming (#2)

This adds the pep8-naming package to the dev requirements and to the pre-commit configuration.

LukeHankey commented:
Have tried with more than just two lines of code and the comment method can sometimes comment out blank lines. Thought I'd add that here in case we want to change that but I don't think it's such a problem that requires fixing currently.

kappq commented:
> Have tried with more than just two lines of code and the comment method can sometimes comment out blank lines. Thought I'd add that here in case we want to change that but I don't think it's such a problem that requires fixing currently.

If you want I can take a look and try to fix it. It shouldn't be too complicated.

kappq committed:
Remove blank line comments

LukeHankey committed:
feat: Add change_function_call_name modifier

LukeHankey committed:
feat: Add insert_empty_statements modifier

LukeHankey commented on modifiers.py:
```diff
@@ -160,6 +160,22 @@ def change_function_call_name(self) -> Self:
+        total_length = len(self.file_contents)
+        random_position = random.randrange(total_length)
+
+        statement = "if kappa_cj9:\n"
```
Taking more suggestions for this as it doesn't always have to be the same string inserted.

kappq replied:
We could try to generate statements randomly, but for now, I would probably just make a list and pick a random one.

LukeHankey replied:
Have added a couple more in [#255e4df ](https://github.com/Vthechamp22/kindly-kappa/pull/7/commits/255e4dfaf5d62a253b5eef4e81246362a5dbbb11) but will add others as they get suggested.

LukeHankey committed:
feat: Add a few statements to randomise the keyword changes and function call names

LukeHankey commented:
I think this PR is ready to be fully reviewed. Will keep as a draft as I am still unsure on how the data will be received.

LukeHankey committed:
fix: random.choice doesn't like to use sets. Changed to a list

LukeHankey commented:
One more thing I'll leave here is that the level of difficulty scales quadratically.

Difficulty | Bugs
1. 1 | 1
2. 2 | 4
3. 3 | 9
etc..

I don' think this should be the intended method of adding code modifications as having 9 bugs can be really difficult to correct as the entire codebase can mark up as red. This will be highlighted even further when adding more bugs every x amount of seconds.
![image](https://user-images.githubusercontent.com/68163698/180603023-fc86d896-53b4-48e1-8481-5afe50dafbe8.png)

LukeHankey committed:
Revert "fix: D101 Missing docstring in public class"

This reverts commit 2c56c617a19792e1b80a96c038a24842b6cf4613.

LukeHankey committed:
docs: Adds class docstring

LukeHankey committed:
Merge branch 'main' into code-modifiers

LukeHankey committed:
Group the modifiers into server/ directory

LukeHankey committed:
feat: Output returns only modified lines of code (line_number, modified_code)

LukeHankey commented on server/modifiers.py:
```diff
@@ -46,7 +47,19 @@ def output(self) -> list[str]:
+                line_diffs.append((int(num), new_value))
+        except KeyError:
+            # No values were changed
+            return [(0, "")]
```
My only worry about this is when sending data back to the client to modify specific lines of code. If no values were changed, this will still return with the first line of code becoming an empty string. Is this acceptable or would it be better to do something else? Perhaps re-run the code again until there is no KeyError or check for this specific value when sending a request back to the client and abort the request if that is the case?

kappq replied:
Couldn't it just return an empty list?

LukeHankey replied:
Yeah, that could work. I think then we would just need a check in place before looping over the output data.

kappq commented on server/modifiers.py:
```diff
@@ -0,0 +1,202 @@
+    test_lines = ["def say_hello() -> str:\n", '    return "Hello!"\n', "say_hello()\n", "\n"]
+
+    modifiers = Modifiers(test_lines)
+    output = modifiers.output
```
It looks like the `output` variable isn't accessed, but performs multiple edits to the final code (I know it's not a bug, it's a feature!).

LukeHankey replied:
Everything in the main guard is just for testing. I probably should have removed it though since I have it printed just below that line.

kappq commented on server/modifiers.py:
```diff
@@ -0,0 +1,202 @@
+import re
+from operator import methodcaller
+
+from deepdiff import DeepDiff
```
The `deepdiff` module should be added to the requirements.

kappq commented on server/modifiers.py:
```diff
@@ -0,0 +1,202 @@
+
+        return self
+
+    def insert_empty_statements(self) -> Self:
```
`insert_empty_statement` inserts multiple newline characters in the line, so I was wondering if they should be returned as multiple lines that were changed, instead of one single line.

If so, the `\n` at the end of every line might not be needed, our editor could just assume that there is a new line between consecutive list elements, which could make the code look a bit cleaner.

For now, I don't think that's an important change, but we might consider it in the future.

LukeHankey replied:
They could be returned as multiple lines, it would just mean that the modified code would be even more buggy, for example a perfectly good line of code would be replaced by a `\n`. I could remove the `\n` from `self.modified_contents[random_position] = f"{self.modified_contents[random_position]}\n{statement}"` just before the statement at the end as I only included it to look better, essentially. If that was removed, then this would only affect one line and not two.

LukeHankey replied:
> If so, the `\n` at the end of every line might not be needed, our editor could just assume that there is a new line between consecutive list elements, which could make the code look a bit cleaner.

I think the newlines need to be included when receiving the data from the client at least. For some reason, the diff doesn't get created without newlines, or just generally without something changing between the input and the output.

For example, try removing the `\n` from the `test_lines` and also the `[:-1]` from the instance attributes in the \_\_init\_\_ and you will always get `[(0, "")]` as the return value. If you can somehow find a way to make this work without relying on any of the modifications from receiving the data, then feel free to PR that into this branch.

kappq commented on server/modifiers.py:
```diff
@@ -0,0 +1,202 @@
+    """A set of code modifying methods."""
+
+    def __init__(self, file_contents: list[str], difficulty: int = 1) -> None:
+        """This class has different functions which introduce different types of bugs.
```
If we want to stick to the [Google code style](https://google.github.io/styleguide/pyguide.html#s3.8.1-comments-in-doc-strings), docstrings shouldn't be longer than 80 characters.

LukeHankey replied:
> If we want to stick to the [Google code style](https://google.github.io/styleguide/pyguide.html#s3.8.1-comments-in-doc-strings), docstrings shouldn't be longer than 80 characters.

Thanks. I would have though the flake8-docstrings should have caught this.

kappq replied:
Yeah, I thought that as well, but apparently not.

LukeHankey replied:
This line has 75 (78 including """), though?

kappq committed:
Add deepdiff to the requirements

LukeHankey committed:
fix: Forgot to add the if keyword as described in the docstring

LukeHankey committed:
fix: return empty list

LukeHankey committed:
fix: remove unreferenced variable

LukeHankey committed:
Merge branch 'main' into code-modifiers
